prov/efa/src/efa_mr.c

```diff
@@ -547,6 +550,139 @@ static struct ibv_mr *efa_mr_reg_ibv_mr(struct efa_mr *efa_mr, struct fi_mr_attr
+	 */
+	if (err == -FI_ENOKEY) {
+		existing_mr_fid = ofi_mr_map_get(&efa_mr->domain->util_domain.mr_map, efa_mr->mr_fid.key);
+		assert(existing_mr_fid);
+		existing_mr = container_of(existing_mr_fid, struct efa_mr, mr_fid);
+
+		if (existing_mr->peer.iface != FI_HMEM_CUDA) {
+			EFA_WARN(FI_LOG_DOMAIN, "key %ld already assigned to buffer: %p hmem_iface: %d length: %ld\n",
+				 existing_mr->mr_fid.key,
+				 existing_mr->ibv_mr->addr,
+				 existing_mr->peer.iface,
+				 existing_mr->ibv_mr->length);
+			return -FI_ENOKEY;
+		}
+
+		assert(existing_mr->peer.iface == FI_HMEM_CUDA);
+		err = ofi_cuPointerGetAttribute(&cuda_memory_type, CU_POINTER_ATTRIBUTE_MEMORY_TYPE,
+						(CUdeviceptr)existing_mr->ibv_mr->addr);
+		if (!err) {
+			EFA_WARN(FI_LOG_DOMAIN, "key %ld already assigned to cuda buffer: %p cuda_memor_type: %d length: %ld\n",
+				 existing_mr->mr_fid.key,
+				 existing_mr->ibv_mr->addr,
+				 cuda_memory_type,
+				 existing_mr->ibv_mr->length);
+			return -FI_ENOKEY;
+		}
+
+		EFA_INFO(FI_LOG_DOMAIN, "key %ld has been assigned to cuda buffer: %p length: %ld, which has since been freed\n",
+			 existing_mr->mr_fid.key,
+			 existing_mr->ibv_mr->addr,
+			 existing_mr->ibv_mr->length);
+
+		assert(efa_mr->domain->cache);
+		pthread_rwlock_rdlock(&mm_list_rwlock);
+		pthread_mutex_lock(&mm_lock);
+		ofi_mr_cache_notify(efa_mr->domain->cache, existing_mr->ibv_mr->addr, existing_mr->ibv_mr->length);
+		pthread_mutex_unlock(&mm_lock);
+		pthread_rwlock_unlock(&mm_list_rwlock);
+
+		/* due to MR cache's deferred de-registration, ofi_mr_cache_notify() only move the region to dead_region_list
+		 * ofi_mr_cache_flush() will actually remove the region from cache.
+		 * lru is a list of regions that are still active, so we set flush_lru to false
+		 */
+		ofi_mr_cache_flush(efa_mr->domain->cache, false /*flush_lru */);
+
+		err = ofi_mr_map_insert(&efa_mr->domain->util_domain.mr_map, mr_attr,
+					&efa_mr->mr_fid.key, &efa_mr->mr_fid);
+	}
+
+	if (err) {
+		EFA_WARN(FI_LOG_MR,
+			"Unable to add MR to map. errno: %d errmsg: (%s) key: %ld buff: %p hmem_iface: %d len: %zu\n",
+			err,
+			fi_strerror(-err),
+			efa_mr->mr_fid.key,
+			mr_attr->mr_iov->iov_base,
+			mr_attr->iface,
+			mr_attr->mr_iov->iov_len);
+		return err;
+	}
+
+	return 0;
+}
+#else
+static
+int efa_mr_update_domain_mr_map(struct efa_mr *efa_mr, struct fi_mr_attr *mr_attr)
+{
+	int err;
+
+	mr_attr->requested_key = efa_mr->mr_fid.key;
+	ofi_genlock_lock(&efa_mr->domain->util_domain.lock);
+	err = ofi_mr_map_insert(&efa_mr->domain->util_domain.mr_map, mr_attr,
+				&efa_mr->mr_fid.key, &efa_mr->mr_fid);
+	if (err) {
+		EFA_WARN(FI_LOG_MR,
+			"Unable to add MR to map. errno: %d errmsg: (%s) key: %ld buff: %p hmem_iface: %d len: %zu\n",
+			err,
+			fi_strerror(-err),
+			efa_mr->mr_fid.key,
+			mr_attr->mr_iov->iov_base,
+			mr_attr->iface,
+			mr_attr->mr_iov->iov_len);
+		return err;
+	}
+
+	return 0;
+}
```

Comment:
need a unlock?

Reply:
unlock of which lock?

Reply:
there is a `ofi_genlock_lock(&efa_mr->domain->util_domain.lock);` before the mr map insert?

Reply:
added